src/neo/SmartContract/ApplicationEngine.cs

```diff
@@ -45,15 +43,21 @@ private class InvocationState
         public UInt160 CurrentScriptHash => CurrentContext?.GetState<ExecutionContextState>().ScriptHash;
         public UInt160 CallingScriptHash => CurrentContext?.GetState<ExecutionContextState>().CallingScriptHash;
         public UInt160 EntryScriptHash => EntryContext?.GetState<ExecutionContextState>().ScriptHash;
-        public IReadOnlyList<NotifyEventArgs> Notifications => notifications;
+        public IReadOnlyList<NotifyEventArgs> Notifications => notifications ?? (IReadOnlyList<NotifyEventArgs>)Array.Empty<NotifyEventArgs>();
 
         public ApplicationEngine(TriggerType trigger, IVerifiable container, StoreView snapshot, long gas, bool testMode = false)
+            : this(trigger, container, snapshot, gas, new ReferenceCounter())
         {
-            this.gas_amount = GasFree + gas;
             this.testMode = testMode;
+        }
+
+        private ApplicationEngine(TriggerType trigger, IVerifiable container, StoreView snapshot, long gas, ReferenceCounter referenceCounter)
```

Comment:
We can't share a referenceCounter under test mode? the wallet will need to replicate the same behaviour.